### PR TITLE
show content to authorized teachers only on dsl defined levels

### PR DESCRIPTION
**Pull request by laurelfan**
- add teacher_markdown property to dsl defined levels
- the content in this property is displayed on level pages to authorized teachers only
- "authorized" teachers are those that are in a cohort
- support encrypting dsl defined levels so this content does not end up in our public git repo

After deploying
- [ ] add encryption key to all environments
- [ ] test creating an encrypted level on levelbuilder and verify that it is seeded correctly in staging

Teacher markdown editor:

Teacher markdown level view:

### CODE REVIEW

File: dashboard/app/controllers/levels_controller.rb [hunk 1]
```diff
@@ -236,8 +232,10 @@ def level_params
       :level_num,
       :user,
       :dsl_text,
+      :encrypted,
       {concept_ids: []},
-      {soft_buttons: []}
+      {soft_buttons: []},
+      {examples: []}
```
Comment (wjordan) [inline dashboard/app/controllers/levels_controller.rb]: is this `examples` parameter important related to this PR somehow, or leftover/typo from something else?

File: dashboard/app/controllers/levels_controller.rb [hunk 2]
```diff
@@ -35,10 +35,6 @@ def edit
     if @level.is_a? Grid
       @level.maze_data = @level.class.unparse_maze(@level.properties)
     end
-    if @level.is_a? DSLDefined
-      @filename = @level.filename
-      @dsl_file = File.read(@filename) if @filename && File.exist?(@filename)
-    end
```
Comment (wjordan) [inline dashboard/app/controllers/levels_controller.rb]: Just to make sure I followed right, this logic was refactored to reuse [dsl_text](https://github.com/code-dot-org/code-dot-org/pull/3397/files#diff-1514f30aa6e8764312feafd5e50f793fR26) instead?

File: dashboard/app/controllers/levels_controller.rb (hunk 1 above)
Comment (laurelfan) [inline dashboard/app/controllers/levels_controller.rb]: yes, leftover from implementing a similar feature (encrypted examples) for applab levels, removing...

File: dashboard/app/controllers/levels_controller.rb (hunk 2 above)
Comment (laurelfan) [inline dashboard/app/controllers/levels_controller.rb]: yes, refactored to move this logic into the model, see also the view template that consumes this: https://github.com/code-dot-org/code-dot-org/pull/3397/files#diff-1514f30aa6e8764312feafd5e50f793f

**Comment by wjordan** [comment]
The code looks good to me, very nice and clean implementation of what could have been a quite complex feature to add into our system. Love the bits of refactoring and added tests.

Not blocking this PR, but I wonder if we absolutely need to hide this content behind a key privately controlled within our server environment, as opposed to making the key available within our public repo. If the goal of this feature is merely to provide a deterrent to devious students from having easy access to answer keys and not a more substantial security issue, I wonder if encrypting with a provided key could provide enough of a deterrent without injecting proprietary content into our codebase.

Otherwise, if this kind of proprietary content becomes a common pattern we end up using often in curriculum development, longer-term I'd like us to consider maintaining a clear distinction between content we make freely available and open-source, and content that we need to explicitly hide from the public for whatever reason, and splitting the latter out into e.g. a Chef data-bag, private Git repository, S3 bucket.

**Comment by laurelfan** [comment]
thanks will! 

I'm going to have to use a different pattern to encrypt properties for the SerializedProperties levels. If we end up doing a refactor to move all levels to some sort of propertybag (json or json in xml) this implementation will go away because dsls will go away.

Either way... I totally agree that it would be nice to have a cleaner separation of 'public' and 'not public' stuff. If we're doing a level format refactor this may be a good time to revisit were/how we store levels in general as well (this whole encryption/hiding thing makes it tricky to do that bundling levels up on a CDN for the client to get though).

**Pull request merged**